plugins/samples/content_injection/plugin.rs

```diff
@@ -0,0 +1,180 @@
+    // otherwise the code will panic and plugin will crash.
+
+    // Stores HTML as the rewriter parses and modifies HTML.
+    // Only stores sections of HTML that have not yet been seen back to client.
+    // After sending data to client, we clear output to avoid unnecessary memory
+    // growth.
+    output: Rc<RefCell<Vec<u8>>>,
+    // HtmlRewriter. Member of MyHttpContext a.k.a "StreamContext" so that the
+    // rewriter persists across multiple body callbacks.The rewriter parses
+    // and modifies chunks of Html. Behavior for Html modifications are defined
+    // via content_handlers. The rewriter only writes completed sections of Html to
+    // the output sink(e.g., when the rewriter parses "<di", nothing will be
+    // sent to the output sink. If the next chunk was "v> <h1>foo</h", the
+    // rewriter would then write "<div><h1>foo" to the output sink) unless
+    // HtmlRewriter::end() is called. HtmlRewriter::end() results in all
+    // uncompleted Html being buffered in rewriter to be sent to output sink as
+    // if it were plain text.
+    rewriter: Option<HtmlRewriter<'a, MyOutputSink>>,
+    // True when plugin has added script to <head>.
+    completed: Rc<RefCell<bool>>,
+}
+
+impl<'a> MyHttpContext<'a> {
+    pub fn new() -> Box<MyHttpContext<'a>> {
+        let output = Rc::new(RefCell::new(Vec::new()));
+        let completed = Rc::new(RefCell::new(false));
+        Box::new(MyHttpContext {
+            output: output.clone(),
+            completed: completed.clone(),
+            rewriter: Some(HtmlRewriter::new(
+                Settings {
+                    element_content_handlers: vec![element!("head", move |el| {
+                        el.prepend(
+                            "<script src=\"https://www.foo.com/api.js\"></script>",
+                            ContentType::Html,
+                        );
+                        *completed.borrow_mut() = true;
+                        Ok(())
+                    })],
+                    ..Settings::new()
+                },
+                MyOutputSink {
+                    output_sink: output,
+                },
+            )),
+        })
+    }
+
+    fn parse_chunk(&mut self, body_bytes: Bytes) -> Result<(), Box<dyn Error>> {
+        // Parse/rewrite current chunk
+        self.rewriter
+            .as_mut()
+            .ok_or("Expected valid rewriter. Got None")?
+            .write(&body_bytes)?;
+        return Ok(());
+    }
+
+    fn end_rewriter(&mut self) -> Result<(), Box<dyn Error>> {
+        // Stop the rewriter after completing desired domain rewrite to
+        // dump any unparsable inputs to output.
+        self.rewriter
+            .take()
+            .expect("Expected valid rewriter. Got None")
+            .end()?;
+        return Ok(());
+    }
+}
+
+impl<'a> Context for MyHttpContext<'a> {}
+
+impl<'a> HttpContext for MyHttpContext<'a> {
+    fn on_http_request_headers(&mut self, _: usize, _: bool) -> Action {
```

Comment:
IIUC, I think this should work without this function implemented at all

Reply:
Removed